File: Ahorn/entities/CustomBadelineBoss.jl
module XaphanHelperCustomBadelineBoss

using ..Ahorn, Maple

@mapdef Entity "XaphanHelper/CustomBadelineBoss" CustomBadelineBoss(x::Integer, y::Integer, patternIndex::Number=1, startHit::Bool=false, cameraPastY::Number=120.0, cameraLockY::Bool=true, canChangeMusic::Bool=true, hitParticleColor1::String="ff00b0", hitParticleColor2::String="ff84d9", shotTrailParticleColor1::String="ffced5", shotTrailParticleColor2::String="ff4f7d", beamDissipateParticleColor::String="e60022", MoveParticleColor1::String="ac3232", MoveParticleColor2::String="e05959", trailColor::String = "ac3232", spriteName::String="badeline_boss")

const placements = Ahorn.PlacementDict(
    "Custom Badeline Boss (Xaphan Helper)" => Ahorn.EntityPlacement(
        CustomBadelineBoss,
    )
)

Ahorn.editingOptions(entity::CustomBadelineBoss) = Dict{String, Any}(
    "patternIndex" => Maple.badeline_boss_shooting_patterns
)

Ahorn.nodeLimits(entity::CustomBadelineBoss) = 0, -1

sprite = "characters/badelineBoss/charge00.png"

function Ahorn.selection(entity::CustomBadelineBoss)
    nodes = get(entity.data, "nodes", ())
    x, y = Ahorn.position(entity)

    res = Ahorn.Rectangle[Ahorn.getSpriteRectangle(sprite, x, y)]
    
    for node in nodes
        nx, ny = Int.(node)

        push!(res, Ahorn.getSpriteRectangle(sprite, nx, ny))
    end

    return res
end

function Ahorn.renderSelectedAbs(ctx::Ahorn.Cairo.CairoContext, entity::CustomBadelineBoss)
    px, py = Ahorn.position(entity)

    for node in get(entity.data, "nodes", ())
        nx, ny = Int.(node)

        Ahorn.drawArrow(ctx, px, py, nx, ny, Ahorn.colors.selection_selected_fc, headLength=6)
        Ahorn.drawSprite(ctx, sprite, nx, ny)

        px, py = nx, ny
    end
end

function Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::CustomBadelineBoss, room::Maple.Room)
    x, y = Ahorn.position(entity)
    Ahorn.drawSprite(ctx, sprite, x, y)
end

end

File: Loenn/entities/CustomBadelineBoss.lua
local enums = require("consts.celeste_enums")

local CustomBadelineBoss = {}

CustomBadelineBoss.name = "XaphanHelper/CustomBadelineBoss"
CustomBadelineBoss.depth = 0
CustomBadelineBoss.nodeLineRenderType = "line"
CustomBadelineBoss.texture = "characters/badelineBoss/charge00"
CustomBadelineBoss.nodeLimits = {0, -1}
CustomBadelineBoss.fieldOrder = {
    "x", "y", "cameraPastY", "patternIndex", "spriteName", "trailColor", "MoveParticleColor1", "MoveParticleColor2", "hitParticleColor1", "hitParticleColor2", "shotTrailParticleColor1", "shotTrailParticleColor2", "beamDissipateParticleColor", "startHit", "cameraLock", "cameraLockY", "canChangeMusic", "drawProjectilesOutline"
}
CustomBadelineBoss.fieldInformation = {
    patternIndex = {
        fieldType = "integer",
        options = enums.badeline_boss_shooting_patterns,
        editable = false
    },
    hitParticleColor1 = {
        fieldType = "color"
    },
    hitParticleColor2 = {
        fieldType = "color"
    },
    shotTrailParticleColor1 = {
        fieldType = "color"
    },
    shotTrailParticleColor2 = {
        fieldType = "color"
    },
    beamDissipateParticleColor = {
        fieldType = "color"
    },
    MoveParticleColor1 = {
        fieldType = "color"
    },
    MoveParticleColor2 = {
        fieldType = "color"
    },
    trailColor = {
        fieldType = "color"
    }
}
CustomBadelineBoss.placements = {
    name = "CustomBadelineBoss",
    data = {
        patternIndex = 1,
        startHit = false,
        cameraPastY = 120.0,
        cameraLockY = true,
        canChangeMusic = true,
        hitParticleColor1 = "ff00b0",
        hitParticleColor2 = "ff84d9",
        shotTrailParticleColor1 = "ffced5",
        shotTrailParticleColor2 = "ff4f7d",
        beamDissipateParticleColor = "e60022",
        MoveParticleColor1 = "ac3232",
        MoveParticleColor2 = "e05959",
        trailColor = "ac3232",
        spriteName = "badeline_boss",
        cameraLock = true,
        drawProjectilesOutline = true
    }
}

return CustomBadelineBoss